chore: lightaccount doc improvement (#153)
alchemyplatform/aa-sdk: site/smart-accounts/accounts/light-account.md

      content: Follow this guide to use Light Account with Account Kit, a vertically integrated stack for building apps that support ERC-4337.
  - - meta
    - name: twitter:title
      content: Light Account Smart Contract
      content: Light Account
  - - meta
    - name: twitter:description
      content: Follow this guide to use Light Account with Account Kit, a vertically integrated stack for building apps that support ERC-4337.
---

# Light Account Smart Contract
# Light Account

> Light Account is a secure, gas-optimized, [ERC-4337](https://eips.ethereum.org/EIPS/eip-4337) smart account implementation.

Reviewer: i'd keep this tbh.

Reviewer 2: it's not gone. down in ## Overview section, first sentence

> **Secure, optimized, and extendable**

Reviewer 3: is it extendable?  it's not eip-6900 compatible so what do we mean by this?

Reviewer 2: meaning forward-compatible with modular account via future upgrade

Reviewer: ah so maybe the keyword was upgradeable?

Reviewer 2: yeah, but upgradeable kind overlapped with the general upgradeable contracts concept, but that's not really what I tried to entail. More of high level umbrella term to imply the overall picture. No strong pref

**Secure, optimized, and extendable**

> We started with the Ethereum Foundation’s canonical [SimpleAccount](https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/samples/SimpleAccount.sol) and added key improvements for production app developers:

Reviewer: i'd also keep this?

Reviewer 2: I feel like people don't really care that the light account is an improvement to SimpleAccount. What's important is that it is efficient, secure, and compatible/portable.

Even web3 devs, most of them don't know about the simple account reference impl by Ethereum, and even those that do know that it is a ref impl, not for production.

Reviewer: sgtm

Reviewer 3: good point, I see both sides here.  I lean toward keeping it for the next ~ 2 months because
* many current AA devs know about SimpleAccount and that is their point of reference
* this builds credibility for us -- we're based on the EF's official version and we improved upon it

i think by early next year this becomes less relevant to the audience and we've built up our brand as a contract developer, then we can remove this

> Light Account is a [ERC-4337](https://eips.ethereum.org/EIPS/eip-4337) smart account solution designed to address the core needs of web3, focusing on efficiency, trust, and security for your account. It turbocharges Ethereum smart accounts by reducing gas costs, and supporting [ERC-1271](https://eips.ethereum.org/EIPS/eip-1271) signatures to ensure compatibility across all decentralized applications (dApps). Additionally, Light Account offers ownership transfer capabilities, ensuring users are not tied to a single signer.

Reviewer 3: ```suggestion
Light Account is an [ERC-4337](https://eips.ethereum.org/EIPS/eip-4337) smart account. It's fully production-ready with a security [audit](https://github.com/alchemyplatform/light-account/blob/main/Quantstamp-Audit.pdf), gas optimizations, and [ERC-1271](https://eips.ethereum.org/EIPS/eip-1271) signature support. Additionally, Light Account support ownership transfer to ensure you and your user don't get locked into a particular signer.
```

## Why Light Account

### Gas-optimized (Benchmarks)

> Light Account is engineered to minimize gas costs, as demonstrated by the benchmark results below. These results highlight its efficiency in various common user flows:

Reviewer 3: ```suggestion
Light Account is engineered to minimize gas costs, as measured by the following benchmarks for common operations like native token transfers, ERC20 transfers, and account creation. 
```

| Alchemy LightAccount                                                                                                    | 100844          | 90345          | 279746   |
| Kernel v2.1-lite                                                                                                        | 101002          | 90321          | 230968   |
| [SimpleAccount](https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/samples/SimpleAccount.sol) | 101319          | 90907          | 383218   |
| Etherspot                                                                                                               | 103719          | 93324          | 279219   |
| Biconomy                                                                                                                | 104408          | 93730          | 270013   |
| Kernel v2.1                                                                                                             | 106460          | 96038          | 265215   |
| Kernel v2.0                                                                                                             | 110018          | 99622          | 339882   |

> ### Secure (Audited and Open-Sourced)

Reviewer 3: ```suggestion
### Secure, Audited, Open Source
```

> Light Account has been thoroughly audited by Quantstamp, ensuring its security and reliability. You can find the audit report [here](https://github.com/alchemyplatform/light-account/blob/main/Quantstamp-Audit.pdf). It's an open-source solution, allowing developers to inspect and validate its code.

Reviewer 3: ```suggestion
Light Account has been audited by Quantstamp. You can find the audit report [here](https://github.com/alchemyplatform/light-account/blob/main/Quantstamp-Audit.pdf). 

Light Account is is also fully open source so you can validate the [source code](https://github.com/alchemyplatform/light-account).
```

### Modular 6900 Accounts (Coming Soon)

> We've designed Light Account with the future in mind, making it forward-compatible and upgradeable for future, more optimized versions and the upcoming modular [ERC-6900](https://eips.ethereum.org/EIPS/eip-6900) account standard, ensuring you'll stay on the cutting edge of EVM accounts.

Reviewer 3: ```suggestion
We're pioneering the future of modular smart accounts with [EIP-6900](https://www.alchemy.com/blog/account-abstraction-erc-6900). We're developing a Modular Account to support EIP-6900 coming soon.

Light Account is designed with that future in mind. It is forward-compatible with EIP-6900 so you can upgrade it to 6900-compatible account once the Modular Account is ready.
```

Reviewer 3: also i wasn't sure how to do this, but can you please link the first instance of 'Modular Account' in this comment to the modular account doc in our repo here?

> ## Using Light Account

Reviewer 3: feels like this should be pulled out into a separate doc. wdyt? feels buried this far down after all the introduction information

Reviewer 2: We already have a separate light account doc section that goes in depth.

# Light Account

### Secure, optimized, and extendable

> In Web3, efficiency, trust, and security are non-negotiable, especially with your account. That's why we built Light Account! With Light Account, we’ve turbocharged Ethereum smart accounts, reducing gas costs and supporting ERC-1271 signatures to add portability amongst all dApps. Plus, we've got your back with ownership transfer, ensuring you're never locked into a single signer. Our benchmark results speak for themselves, giving you the confidence to use a fully audited, optimized, and [open-sourced](https://github.com/alchemyplatform/light-account) smart account solution.

Reviewer 3: this paragraph is trying to do too many things at once. Let's structure this doc into discrete sections for clarity
* Overview:  it's important for docs to start at square one. First section should simply define what Light Account is. I think we should keep the section that was here previously, and then add what you wrote as a "Why Light Account" section explaining the benefits. 
* Why Light Account
* * Gas-optimized (benchmarks)
* * Secure (audited, etc)
* * Coming soon: Modular

Copy and voice notes:
* web3 lowercase
* turbocharged is hyperbolic
* 'we've got your back" is a bit too casual / personifying the contract
* "Our benchmark results speak for themselves, giving you the confidence" also too casual. Goal is to follow the [Google style guidelines](https://developers.google.com/style/tone)

### Future-Proofed with Modular 6900 Accounts:

We also built Light Account with the future in mind. Light Account is forward-compatible and upgradeable for future, more optimized versions and the upcoming modular 6900 account standard, ensuring you'll stay on the cutting edge of EVM accounts. 

> ### The Responsible Choice

Reviewer 3: I think this is accidentally overstating. It almost reads as if it's irresponsible to choose other smart accounts, which isn't true (any audited contract is responsible)

> At Alchemy, in partnership with the community, we focus on the fundamentals and build our solutions from the ground up. Your trust and safety are paramount in the new digital frontier, so ensure you and your users are backed by a thoughtful craft and standard. The future of accounts is within the Alchemy Family of Smart Accounts - Optimized, Open, and Secure.

Reviewer 3: the first two sentences sound fluffy. can we be more concrete?

the last sentence
* style: Alchemy is the only word that should be capitalized there
* do we have a family? that's a pretty expansive concept that I'm not sure we're building. Light + Modular are two options but it's not exactly a family. Let's tone this down and be more concrete

> ## Using Light Account

Reviewer 3: in my opinion this doc is getting long (which is good!) and for clarity we should split them up into two separate docs:
* Why Light Account
* Using Light Account

left more details in the [asana ticket](https://app.asana.com/0/1205598834384806/1205608436455254/f).  This doesn't have to be a blocker for this PR, we can do it later if easier

The code snippet below demonstrates the usage of Light Account with Account Kit. It creates a Light Account and sends a `UserOperation` from it:

<!--@include: ../../getting-started.md{56,68}-->

> ## Details

Reviewer 3: this is a catch all section that won't be very useful to a reader, or at least make them do the work to find the info they need. it would be better to cover this in an appropriate section above and delete this section

will leave suggestions of how to restructure this

> Light Account is [audited by quantstamp](https://github.com/alchemyplatform/light-account/blob/main/Quantstamp-Audit.pdf), is gas-optimized, and compatible as a [ERC-4337](https://eips.ethereum.org/EIPS/eip-4337) smart account implementation.

Reviewer 3: great definition of what it is. let's put in the Overview section

> Based on Ethereum Foundation’s canonical [SimpleAccount](https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/samples/SimpleAccount.sol) with key improvements for production app developers, Light Account:

Reviewer 3: Overview

> - significantly reduce gas costs

Reviewer 3: these could go in overview or Why Light Account?  or perhaps a "Features" section makes sense if those get long

- [ERC-1271](https://eips.ethereum.org/EIPS/eip-1271) has signature support to ensure users can sign messages, such as on OpenSea
- allows ownership transfer so that users won’t get locked into a single `Signer`

> ### Benchmarks

Reviewer 3: let's put this in the "gas-optimized" section of Why Light Account

# Light Account Smart Contract

### Secure, optimized, and extendable

## Overview

Light Account is a secure, gas-optimized, and versatile Ethereum smart account solution. It introduces efficiency, trust, and security to your web3 experience. In this document, we'll explore why Light Account is a valuable choice for developers.

## Why Light Account

> Light Account is designed to meet your web3 needs, offering the following benefits:

Author: "designed to meet your web3 needs", it sounds to high level. Can we be more specific? 

"is a fully functioning web3 account, via smart contracts. It offers the following benefits"

At Alchemy, we prioritize fundamental principles and community collaboration. Your trust and safety are paramount in the new digital frontier. Choose Light Account for an optimized, open, and secure smart account solution.

## Using Light Account

The code snippet below demonstrates how to use Light Account with Account Kit. It creates a Light Account and sends a `UserOperation` from it:

<!--@include: ../../getting-started.md{56,68}-->

## Details

Light Account is [audited by quantstamp](https://github.com/alchemyplatform/light-account/blob/main/Quantstamp-Audit.pdf), is gas-optimized, and compatible as a [ERC-4337](https://eips.ethereum.org/EIPS/eip-4337) smart account implementation.

Based on Ethereum Foundation’s canonical [SimpleAccount](https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/samples/SimpleAccount.sol) with key improvements for production app developers, Light Account:

- significantly reduce gas costs
- [ERC-1271](https://eips.ethereum.org/EIPS/eip-1271) has signature support to ensure users can sign messages, such as on OpenSea
- allows ownership transfer so that users won’t get locked into a single `Signer`

## Benchmarks

> Light Account was built to optimize the most common user flows, providing the lowest gas cost for transfers. Here is a table of Light Account benchmarked against other smart account implementations:

Author: We can probably shorten this.

"Light Account optimizes for transfers (the most common user flow). Below is a table of Light Account benchmark against other smart account implementations:"
